Collective dispatching from Process Group

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91257

Fixes https://github.com/pytorch/pytorch/issues/90932
Fixes https://github.com/pytorch/pytorch/issues/90659

Remove redundant collection operation definitions by calling the ops directly from `ProcessGroup`

Context:
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D42854676](https://our.internmc.facebook.com/intern/diff/D42854676)

H-Huang committed:
Collective dispatching from Process Group

[ghstack-poisoned]

H-Huang committed:
Update on "Collective dispatching from Process Group"


Fixes https://github.com/pytorch/pytorch/issues/90932
Fixes https://github.com/pytorch/pytorch/issues/90659

Remove redundant collection operation definitions by calling the ops directly from `ProcessGroup`


[ghstack-poisoned]

H-Huang committed:
Update on "Collective dispatching from Process Group"


Fixes https://github.com/pytorch/pytorch/issues/90932
Fixes https://github.com/pytorch/pytorch/issues/90659

Remove redundant collection operation definitions by calling the ops directly from `ProcessGroup`


[ghstack-poisoned]

H-Huang committed:
Update on "Collective dispatching from Process Group"


Fixes https://github.com/pytorch/pytorch/issues/90932
Fixes https://github.com/pytorch/pytorch/issues/90659

Remove redundant collection operation definitions by calling the ops directly from `ProcessGroup`


[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42854676).

kwen2501 commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -58,387 +58,6 @@ TORCH_LIBRARY(c10d, m) {
-    int64_t dstRank,
-    int64_t tag) {
-  static auto op = c10::Dispatcher::singleton()
-                       .findSchemaOrThrow("c10d::send", "")
-                       .typed<c10::intrusive_ptr<::c10d::Work>(
-                           at::TensorList,
-                           const c10::intrusive_ptr<::c10d::ProcessGroup>&,
-                           int64_t,
-                           int64_t)>();
-  return op.call(tensors, process_group, dstRank, tag);
-}
-
-c10::intrusive_ptr<Work> recv(
-    const c10::intrusive_ptr<ProcessGroup>& process_group,
-    at::TensorList tensors,
-    int64_t srcRank,
-    int64_t tag) {
-  static auto op = c10::Dispatcher::singleton()
-                       .findSchemaOrThrow("c10d::recv_", "")
-                       .typed<c10::intrusive_ptr<::c10d::Work>(
-                           at::TensorList,
-                           const c10::intrusive_ptr<::c10d::ProcessGroup>&,
-                           int64_t,
-                           int64_t)>();
-  return op.call(tensors, process_group, srcRank, tag);
-}
-
-c10::intrusive_ptr<Work> recv_any_source(
-    const c10::intrusive_ptr<ProcessGroup>& process_group,
-    at::TensorList tensors,
-    int64_t tag) {
-  static auto op = c10::Dispatcher::singleton()
-                       .findSchemaOrThrow("c10d::recv_any_source_", "")
-                       .typed<c10::intrusive_ptr<::c10d::Work>(
-                           at::TensorList,
-                           const c10::intrusive_ptr<::c10d::ProcessGroup>&,
-                           int64_t)>();
-  return op.call(tensors, process_group, tag);
-}
-
```
After removing this code block, should we also remove the corresponding API declarations in Ops.hpp?

H-Huang replied:
Good point! Thanks

kwen2501 reviewed (approved):
Thanks for the clean-up! LGTM.
Please see my inline comment.

H-Huang committed:
Update on "Collective dispatching from Process Group"


Fixes https://github.com/pytorch/pytorch/issues/90932
Fixes https://github.com/pytorch/pytorch/issues/90659

Remove redundant collection operation definitions by calling the ops directly from `ProcessGroup`

Context:
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D42854676](https://our.internmc.facebook.com/intern/diff/D42854676)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42854676).

H-Huang commented:
@pytorchbot merge
